### PR TITLE
Use IPLD Dag instead of CoreAPI

### PR DESCRIPTION
**Pull request by Wondertan**
The PR's main goal is to facilitate everything related to IPFS usage by:
1. Using IPLD Merkle DAG interfaces instead of CoreAPI.
2. Reworking Data retrieval logic to be based solely on IPLD Dag(optimization)
3. Use of Dag Sessions(optimization)
4. Reworking and cleaning some tests

DAS timings before 2 above optimizations:
```
I[2021-05-23|23:12:11.531] Successfully finished DAS sampling           height=629 numSamples=15 elapsedtime=1.015415079s
I[2021-05-23|23:12:11.768] Starting Data Availability sampling          height=630 numSamples=15 squareWidth=16
I[2021-05-23|23:12:12.690] Successfully finished DAS sampling           height=630 numSamples=15 elapsedtime=921.977083ms
I[2021-05-23|23:12:13.060] Starting Data Availability sampling          height=631 numSamples=15 squareWidth=16
I[2021-05-23|23:12:14.801] Successfully finished DAS sampling           height=631 numSamples=15 elapsedtime=1.740647283s
I[2021-05-23|23:12:15.003] Starting Data Availability sampling          height=632 numSamples=15 squareWidth=16
```

DAS timings after optimizations:
```
I[2021-05-23|23:13:44.353] Successfully finished DAS sampling           height=648 numSamples=15 elapsedtime=239.112542ms
I[2021-05-23|23:13:44.485] Starting Data Availability sampling          height=649 numSamples=15 squareWidth=16
I[2021-05-23|23:13:44.719] Successfully finished DAS sampling           height=649 numSamples=15 elapsedtime=234.11647ms
I[2021-05-23|23:13:44.853] Starting Data Availability sampling          height=650 numSamples=15 squareWidth=16
I[2021-05-23|23:13:45.095] Successfully finished DAS sampling           height=650 numSamples=15 elapsedtime=242.104799ms
I[2021-05-23|23:13:45.227] Starting Data Availability sampling          height=651 numSamples=15 squareWidth=16
```

NOTE: Heights in the run with optimization was not DASed before, so light-client didn't have common data before.

### CODE REVIEW

File: p2p/ipld/read.go [hunk 1]
```diff
@@ -224,25 +224,35 @@ func GetLeafData(
...
+	if len(lnks) == 1 {
+		return dag.Get(ctx, lnks[0].Cid)
+	}
+
+	total /= 2
+	if leaf < total {
+		root = lnks[0].Cid
+	} else {
+		root, leaf = lnks[1].Cid, leaf-total
+	}
```
Comment (evan-forbes) [inline p2p/ipld/read.go:253]: mind adding some documentation somewhere? doesn't have to be anything too elaborate, but I think some simple guidance would make a big difference. Particularly for readers who are not familiar with our plugin.

Comment (Wondertan) [reply p2p/ipld/read.go:253]: Sure. Also, that’s no the final form. I think to close this PR actually and rewrite it afresh to be based on the master, but that requires #323 to be merged first.

Comment (liamsi) [reply p2p/ipld/read.go:253]: merged :-)

**Comment by liamsi** [comment]
This looks good. I don't fully understand what makes the big difference in latency possible (yet). Will do a full review after https://github.com/lazyledger/lazyledger-core/pull/352#discussion_r637774071

**Comment by Wondertan** [comment]
@liamsi, @evan-forbes, I decided not to close this PR and instead extend it with the requested changes for CoreAPI removal

**Comment by Wondertan** [comment]
In follow-up PR I will go even further and will propose our custom interface for network DataAvailavility which works over DAG

File: consensus/byzantine_test.go [hunk 2]
```diff
@@ -9,6 +9,7 @@ import (
 	"testing"
 	"time"
 
+	mdutils "github.com/ipfs/go-merkledag/test"
```
Comment (liamsi) [inline consensus/byzantine_test.go:12]: Can we use another alias? mdutils sounds like markdown utils. Suggestion: dagutils?

File: light/client.go [hunk 3]
```diff
@@ -7,6 +7,8 @@ import (
 	"fmt"
 	"time"
 
+	format "github.com/ipfs/go-ipld-format"
+	dag "github.com/ipfs/go-merkledag"
```
Comment (liamsi) [inline light/client.go:11]: why even alias this? This will shadow the variable `dag` and `merkledag` is more descriptive (and would be the default without that alias).

File: p2p/ipld/read.go [hunk 4]
```diff
@@ -27,7 +25,7 @@ var ErrTimeout = fmt.Errorf("%s %s", baseErrorMsg, "timeout")
 func RetrieveBlockData(
 	ctx context.Context,
 	dah *types.DataAvailabilityHeader,
-	api coreiface.CoreAPI,
+	dag format.NodeGetter,
```
Comment (liamsi) [inline p2p/ipld/read.go:28]:
```suggestion
	nodeGetter format.NodeGetter,
```
?

File: p2p/ipld/read.go (hunk 1 above)
Comment (Wondertan) [reply p2p/ipld/read.go]: I added some documentation now, hope it will help understand the logic

File: p2p/ipld/read_test.go [hunk 5]
```diff
@@ -8,172 +8,68 @@ import (
...
-		{
-			input:    16,
-			expected: true,
-		},
-	}
-	for _, tt := range tests {
-		res := isPowerOf2(tt.input)
-		assert.Equal(t, tt.expected, res, fmt.Sprintf("input was %d", tt.input))
-	}
-}
```
Comment (liamsi) [inline p2p/ipld/read_test.go:98]: Did you delete these two tests because they test private helpers?

File: p2p/ipld/read_test.go [hunk 6]
```diff
@@ -227,70 +120,55 @@ func TestRetrieveBlockData(t *testing.T) {
...
 
 	for _, tc := range tests {
 		tc := tc
-
 		t.Run(fmt.Sprintf("%s size %d", tc.name, tc.squareSize), func(t *testing.T) {
-			// if we're using the race detector, skip some large tests due to time and
-			// concurrency constraints
-			if racedetector.IsActive() && tc.squareSize > 8 {
-				t.Skip("Not running large test due to time and concurrency constraints while race detector is active.")
-			}
```
Comment (liamsi) [inline p2p/ipld/read_test.go:253]: Can we delete the racedetector package then, or is it still used somewhere else?

**Review by liamsi — COMMENTED** [review]
This looks great! Left some suggestions and questions.

File: p2p/ipld/read.go (hunk 1 above)
Comment (liamsi) [reply p2p/ipld/read.go]: Cool! Just saw that after reviewing.
> Use of Dag Sessions(optimization)

I think I understand this now. Although it's not clear to me how it works under the hood. But it makes sense intuitively.

File: p2p/ipld/read_test.go (hunk 6 above)
Comment (Wondertan) [reply p2p/ipld/read_test.go:253]: I thought I deleted it, but it got lost in stash

Comment (Wondertan) [reply p2p/ipld/read_test.go:253]: Deleting it

File: p2p/ipld/read.go [hunk 7]
```diff
@@ -224,53 +222,40 @@ func GetLeafData(
...
+	return nd.RawData()[1:], nil
+}
 
-	// resolve the path
-	node, err := api.ResolveNode(ctx, p)
+// GetLeafData fetches and returns the raw leaf.
+// It walks down the IPLD NMT tree until it finds the requested one.
+func GetLeaf(ctx context.Context, dag format.NodeGetter, root cid.Cid, leaf, total uint32) (format.Node, error) {
+	// request the node
+	nd, err := dag.Get(ctx, root)
```
Comment (evan-forbes) [inline p2p/ipld/read.go:239]: the new `GetLeaf` is much simpler! I like it a lot :+1:

File: p2p/ipld/read_test.go (hunk 5 above)
Comment (Wondertan) [reply p2p/ipld/read_test.go:98]: I delete tests as I deleted those helpers - there are no more needed.

File: light/client.go (hunk 3 above)
Comment (Wondertan) [reply light/client.go:11]: I think IDE did that, will change in something more decriptive

File: consensus/byzantine_test.go (hunk 2 above)
Comment (Wondertan) [reply consensus/byzantine_test.go:12]: That's a default package name.

Comment (Wondertan) [reply consensus/byzantine_test.go:12]: I won't change that as that's actually temporary thing

File: p2p/ipld/read_test.go (hunk 5 above)
Comment (liamsi) [reply p2p/ipld/read_test.go:98]: Oh, I see. NVM.

File: consensus/byzantine_test.go (hunk 2 above)
Comment (Wondertan) [reply consensus/byzantine_test.go:12]: Later in block propagation PR instead of just Mock I will use Mock with Mock networking

Comment (Wondertan) [reply consensus/byzantine_test.go:12]: So I won't spend time for this renaming

Comment (liamsi) [reply consensus/byzantine_test.go:12]: > That's a default package name.

But what does that even mean? Is the package is aliased like this in other contexts?

File: p2p/ipld/read.go (hunk 4 above)
Comment (Wondertan) [reply p2p/ipld/read.go:28]: Naaah, too long. I don't understand why it is more "descriptive" to repeat words:
>  nodeGetter format.NodeGetter

Here we use words "node" and "getter" two times and I try to avoid that

**Comment by liamsi** [comment]
> race: limit on 8128 simultaneously alive goroutines is exceeded, dying
> FAIL	github.com/lazyledger/lazyledger-core/p2p/ipld	16.744s

This is why the race detector hack was there btw.

File: p2p/ipld/read.go (hunk 4 above)
Comment (liamsi) [reply p2p/ipld/read.go]: OK

**Comment by Wondertan** [comment]
@liamsi, I guessed that Mock IPFS will decrease the number of goroutines to remove the hack. WIll try something else to fix that

File: light/client.go (hunk 3 above)
Comment (Wondertan) [reply light/client.go]: Removed the alias

File: p2p/ipld/read.go (hunk 7 above)
Comment (evan-forbes) [reply p2p/ipld/read.go]: @liamsi do we need to update ADR002 with this slight api change?

File: p2p/ipld/read_test.go [hunk 8]
```diff
@@ -227,70 +120,55 @@ func TestRetrieveBlockData(t *testing.T) {
...
 		{"max square size", consts.MaxSquareSize, false, ""},
 	}
 
 	for _, tc := range tests {
 		tc := tc
-
 		t.Run(fmt.Sprintf("%s size %d", tc.name, tc.squareSize), func(t *testing.T) {
-			// if we're using the race detector, skip some large tests due to time and
-			// concurrency constraints
-			if racedetector.IsActive() && tc.squareSize > 8 {
```
Comment (liamsi) [inline p2p/ipld/read_test.go:251]: To save you some time and to unblock this PR: we can either revert the changes regarding the racedetector, or, we simply skip all tests with a `squareSize > 8`.

File: p2p/ipld/read.go (hunk 7 above)
Comment (liamsi) [reply p2p/ipld/read.go]: Yeah, it would be good if the ADR would match the implementation's APIs.

**Review by evan-forbes — APPROVED** [review]
two :+1: from me! all of these refactors and optimizations are really adding up! :sparkles:

**Comment by Wondertan** [comment]
@liamsi, @evan-forbes, after looking deeply into the reasons why the issue above happens, I conclude that we have a goroutine leak that we should fix. With that hack, we just solely ignored the issue, so I don't think taking the hack back makes sense.

**Comment by Wondertan** [comment]
I would like to handle this myself as part of this PR, but I am really tired to do that today and it is midnight my time

**Comment by evan-forbes** [comment]
we don't have to merge this tonight. I'll try limiting the goroutines with this PR and see what happens

**Comment by Wondertan** [comment]
Ok, but I already pushed temporary hack to unblock this

**Comment by liamsi** [comment]
While it makes perfect sense to limit that number of go-routines spawned in that method, it's not a go-routine leak. Also note that the error only kicks in when the race-detector is active and the code works fine even with larger blocks without the race detector though.

I think, we can skip that test for now for larger blocks with a note that we should limit the number of go-routines spawned (I'll open an issue). If possible, let's merge this and handle the issue in a separate PR.

File: p2p/ipld/read_test.go [hunk 9]
```diff
@@ -227,70 +120,60 @@ func TestRetrieveBlockData(t *testing.T) {
...
+		{"16 KB block timeout expected", 8, true, "not found"},
 		{"max square size", consts.MaxSquareSize, false, ""},
 	}
 
 	for _, tc := range tests {
-		tc := tc
+		// TODO(Wondertan): remove this
+		if tc.squareSize > 8 {
+			continue
+		}
```
Comment (liamsi) [inline p2p/ipld/read_test.go:135]:
```suggestion
		if tc.squareSize > 8 {
			t.Skip("TODO: limit the number of spawned go-routines otherwise we can't run this test while race detector is active.")
		}
```

Comment (Wondertan) [reply p2p/ipld/read_test.go:135]: This wont work as skip has to be run inside t.Run

Comment (liamsi) [reply p2p/ipld/read_test.go:135]: Ah I see. Thought this was still inside the Run. it's better to log that we are skipping that case instead of silently continuing, no?

**Comment by Wondertan** [comment]
> it's not a go-routine leak
That's true! Goruitne leak is when goroutine hangs out infinitely without returning, e.g. not calling cancel() for context will introduce one.

My original intuition was that spawning tonnes of routines are like leaking something. Obviously, it's wrong, affects performance, and should be avoided. Idk how to call that properly btw.

File: p2p/ipld/read_test.go (hunk 9 above)
Comment (Wondertan) [reply p2p/ipld/read_test.go:135]: We can keep this as it is. As me or Even will fix that in follow-up PRs.

Comment (liamsi) [reply p2p/ipld/read_test.go:135]: https://github.com/lazyledger/lazyledger-core/issues/357

File: p2p/ipld/read.go (hunk 7 above)
Comment (Wondertan) [reply p2p/ipld/read.go]: I think so, but also I am going to propose an elegant PR soon. If it's get accepted, updating this will make more sense

**Review by liamsi — APPROVED** [review]
Awesome work! 👍🏼 🚀
